### PR TITLE
feat(lora): add de-quantized support for fuse.py

### PR DESCRIPTION
**Pull request by mzbac**
Quick and dirty implementation for supporting de-quant model during fusing. Also, updated the model weight to use HF format so that once the model is fused, it can be converted by gguf for more advanced quant support. 

PS: Once the Lora example is merged into mlx-lm, we can properly add this functionality and make the fused model can be directly loaded by the transformers library to support AWQ/GPTQ.

### CODE REVIEW

File: lora/utils.py [hunk 1]
```diff
@@ -112,15 +112,27 @@ def make_shards(weights: dict, max_file_size_gibibyte: int = 15):
...
-def save_model(save_dir: str, weights, tokenizer, config):
+def save_model(save_dir: str, weights, tokenizer, config, de_quantize: bool = False):
     save_dir = Path(save_dir)
     save_dir.mkdir(parents=True, exist_ok=True)
-    shards = make_shards(weights)
+
+    shards = make_shards(weights, max_file_size_gibibyte=5)
+    shards_count = len(shards)
+    shard_file_format = (
+        "model-{:05d}-of-{:05d}.safetensors"
```
Comment (awni) [inline lora/utils.py:122]: It was a bad idea to not use this format in `mlx-lm` as well. We should update it there as well

File: lora/utils.py [hunk 2]
```diff
@@ -112,15 +112,27 @@ def make_shards(weights: dict, max_file_size_gibibyte: int = 15):
     return shards
 
 
-def save_model(save_dir: str, weights, tokenizer, config):
+def save_model(save_dir: str, weights, tokenizer, config, de_quantize: bool = False):
     save_dir = Path(save_dir)
     save_dir.mkdir(parents=True, exist_ok=True)
-    shards = make_shards(weights)
+
+    shards = make_shards(weights, max_file_size_gibibyte=5)
```
Comment (awni) [inline lora/utils.py:119]: Curious, why did you change the max size here?

**Review by awni — COMMENTED** [review]
Looks great, thanks for adding that option! That will be super useful! Just a minor comment. Could you address and then we can merge?

File: lora/utils.py (hunk 1 above)
Comment (mzbac) [reply lora/utils.py:122]: I updated it in lora PR https://github.com/ml-explore/mlx-examples/pull/337

File: lora/utils.py (hunk 2 above)
Comment (mzbac) [reply lora/utils.py:119]: > Curious, why did you change the max size here?

just try to it easier to upload. Otherwise, uploading 15GB takes too long and if anything happens during the upload, users need to re-upload. and, 5GB seems to be the most common size in the Huggingface repo.

File: lora/fuse.py [hunk 3]
```diff
@@ -67,8 +74,33 @@
...
+                    linear.bias = m.bias
+                de_quantize_layers.append((n, linear))
+
+        model.update_modules(tree_unflatten(de_quantize_layers))
+
     weights = dict(tree_flatten(model.parameters()))
-    utils.save_model(args.save_path, weights, tokenizer, config)
+    utils.save_model(
+        args.save_path, weights, tokenizer, config, de_quantize=args.de_quantize
+    )
```
[the remark this answers is not in the document]
Comment (mzbac) [reply lora/fuse.py]: updated: )

**Review by awni — APPROVED** [review]
Thank you!
